### PR TITLE
Use /tmp instead of /dev/shm for object store on Linux if /dev/shm is too small.

### PR DESCRIPTION
**Pull request by robertnishihara**
This should allow Ray to work out of the box on kaggle or in docker containers by falling back to `/tmp` instead of `/dev/shm` when the shared memory file system is too small.

This also changes the default object store size on Mac (80% -> 40%) of total memory and Linux (45% -> 40%). Mostly because on MacOS things seem to get really slow when the object store fills up.

This is not without a cost, if we do end up defaulting to `/tmp` it can be much slower.

EDIT: For example, on a smallish machine, using `/dev/shm`, putting a 1GB object in the object store takes about 300ms. Using `/tmp` it sometimes takes 300ms, and sometimes takes up to 7s.

### CODE REVIEW

**Comment by robertnishihara** [comment]
This replaces #2459.

cc @richardliaw @pcmoritz @atumanov

File: python/ray/services.py [hunk 1]
```diff
@@ -957,12 +957,72 @@ def start_raylet(redis_address,
...
 
+def determine_plasma_store_config(object_store_memory=None,
+                                  plasma_directory=None,
+                                  huge_pages=False):
+    """Figure out how to configure the plasma object store.
+
+    Args:
+        object_store_memory: The user-specified object store memory parameter.
+        plasma_directory: The user-specified plasma directory parameter.
+        huge_pages: The user-specified huge pages parameter.
```
Comment (richardliaw) [inline python/ray/services.py]: Types?

File: python/ray/utils.py [hunk 2]
```diff
@@ -324,6 +324,28 @@ def get_system_memory():
...
 
 
+def get_shared_memory_bytes():
+    """Get the size of the shared memory file system.
+
+    Returns:
+        The size of the shared memory file system in bytes.
+    """
+    # Make sure this is only called on Linux.
+    assert sys.platform == "linux" or sys.platform == "linux2"
```
Comment (richardliaw) [inline python/ray/utils.py]: nit: `assert sys.platform in {"linux", "linux2"}`

File: python/ray/services.py [hunk 3]
```diff
@@ -957,12 +957,72 @@ def start_raylet(redis_address,
...
+    # Determine which directory to use. By default, use /tmp on MacOS and
+    # /dev/shm on Linux, unless the shared-memory file system is too small,
+    # in which case we default to /tmp on Linux.
+    if plasma_directory is None:
+        if sys.platform == "linux" or sys.platform == "linux2":
+            shm_avail = ray.utils.get_shared_memory_bytes()
+            # Compare the requested memory size to the memory available in
+            # /dev/shm.
+            if shm_avail > object_store_memory:
+                plasma_directory = "/dev/shm"
```
Comment (richardliaw) [inline python/ray/services.py]: Can you add a log statement here?

File: python/ray/services.py [hunk 4]
```diff
@@ -957,12 +957,72 @@ def start_raylet(redis_address,
...
+                    "WARNING: The object store is using /tmp instead of "
+                    "/dev/shm because /dev/shm has only {} bytes available. "
+                    "This may slow down performance! You may be able to free "
+                    "up space by deleting files in /dev/shm or terminating "
+                    "any running plasma_store_server processes. If you are "
+                    "inside a Docker container, you may need to pass an "
+                    "argument with the flag '--shm-size' to 'docker run'."
+                    .format(shm_avail))
+        else:
+            plasma_directory = "/tmp"
```
Comment (richardliaw) [inline python/ray/services.py]: wait this will default to `/tmp` for mac?

File: python/ray/services.py [hunk 5]
```diff
@@ -957,12 +957,72 @@ def start_raylet(redis_address,
     return raylet_name
 
 
+def determine_plasma_store_config(object_store_memory=None,
+                                  plasma_directory=None,
+                                  huge_pages=False):
+    """Figure out how to configure the plasma object store.
+
```
Comment (richardliaw) [inline python/ray/services.py]: Can this docstring provide some more insight into what is going on?

File: python/ray/services.py (hunk 4 above)
Comment (robertnishihara) [reply python/ray/services.py]: Yeah, that is the behavior before and after this PR. `/dev/shm` doesn't exist on MacOS I think.

File: python/ray/services.py (hunk 3 above)
Comment (robertnishihara) [reply python/ray/services.py]: What do you want it to say? The directory? I could add that in `start_plasma_store`.

Comment (robertnishihara) [reply python/ray/services.py]: Added there.

File: python/ray/services.py (hunk 5 above)
Comment (robertnishihara) [reply python/ray/services.py]: Done

File: python/ray/services.py (hunk 1 above)
Comment (robertnishihara) [reply python/ray/services.py]: Fixed. However, we haven't been consistent about types. Do you think we should use them or not? Do they add value?

**Review by richardliaw — APPROVED** [review]

**Pull request merged**